File: bot.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
	"strings"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
)

const openAIURL = "https://api.openai.com/v1/engines/davinci/completions"

func chatWithGPT(userMessage string) string {
	reqBody := strings.NewReader(`{
		"prompt": "` + userMessage + `",
		"max_tokens": 150
	}`)

	req, err := http.NewRequest("POST", openAIURL, reqBody)
	if err != nil {
		return "Ошибка: " + err.Error()
	}

	req.Header.Set("Authorization", "Bearer "+os.Getenv("OPENAI_API_KEY"))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "Ошибка: " + err.Error()
	}
	defer resp.Body.Close()

	var responseJson map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&responseJson)

	choices := responseJson["choices"].([]interface{})
	if len(choices) > 0 {
		return choices[0].(map[string]interface{})["text"].(string)
	}

	return "Извините, я не могу обработать ваш запрос."
}

func main() {
	botToken := os.Getenv("TELEGRAM_BOT_TOKEN")
	allowedUserIDs := strings.Split(os.Getenv("ALLOWED_USER_IDS"), ",")

	bot, err := tgbotapi.NewBotAPI(botToken)
	if err != nil {
		log.Panic(err)
	}

	log.Printf("Authorized on account %s", bot.Self.UserName)

	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates, err := bot.GetUpdatesChan(u)

	for update := range updates {
		if update.Message == nil {
			continue
		}

		userID := update.Message.From.ID
		isAllowed := false
		for _, allowedID := range allowedUserIDs {
			if allowedID == userID {
				isAllowed = true
				break
			}
		}

		if !isAllowed {
			bot.Send(tgbotapi.NewMessage(update.Message.Chat.ID, "Извините, у вас нет доступа к этому боту."))
			continue
		}

		response := chatWithGPT(update.Message.Text)
		bot.Send(tgbotapi.NewMessage(update.Message.Chat.ID, response))
	}
}

File: bot.py
import requests
import os
from telegram import Bot, Update
from telegram.ext import MessageHandler, Filters, CallbackContext, Updater

# Загрузка переменных окружения из файла .env
TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
ALLOWED_USER_IDS = os.getenv("ALLOWED_USER_IDS").split(',')

def chat_with_gpt(user_message: str) -> str:
    headers = {
        'Authorization': f'Bearer {OPENAI_API_KEY}',
        'Content-Type': 'application/json'
    }
    data = {
        'prompt': user_message,
        'max_tokens': 150
    }
    response = requests.post('https://api.openai.com/v1/engines/davinci/completions', headers=headers, json=data)
    response_json = response.json()

    return response_json['choices'][0]['text'].strip()

def handle_text(update: Update, context: CallbackContext) -> None:
    user_id = str(update.message.from_user.id)
    
    if user_id not in ALLOWED_USER_IDS:
        update.message.reply_text("Извините, у вас нет доступа к этому боту.")
        return

    user_message = update.message.text
    gpt_response = chat_with_gpt(user_message)
    update.message.reply_text(gpt_response)

def main():
    updater = Updater(token=TELEGRAM_BOT_TOKEN)
    dp = updater.dispatcher
    dp.add_handler(MessageHandler(Filters.text & ~Filters.command, handle_text))
    updater.start_polling()
    updater.idle()

if __name__ == "__main__":
    main()
